File: backend/app/services/auth_service.py
"""User persistence, authentication, and session bookkeeping.

Hashing and token issuance live in `app.core.security` — this module
orchestrates them with the database.

The previous version used raw SHA-256 for passwords and issued opaque
SHA-256 access tokens. Both are gone; passwords now use bcrypt and tokens
are signed JWTs (see `app.core.security`).
"""

from __future__ import annotations

import uuid
from datetime import datetime, timedelta, timezone
from typing import Optional

from sqlalchemy import select, delete
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import settings
from app.core.security import (
    create_access_token,
    create_refresh_token,
    hash_password,
    verify_password,
)
from app.models import User, UserSession, PasswordResetToken
from app.schemas.auth import UserCreate
from app.services.email_service import generate_reset_code, send_password_reset_email


def _utcnow_naive() -> datetime:
    """Naive UTC "now", safe to compare against naive DateTime columns.

    Postgres stores our `timestamp` columns without time zone; asyncpg /
    SQLAlchemy sometimes returns them with tzinfo=UTC on the Vercel
    runtime. Comparing those against `datetime.utcnow()` (naive) raises
    `TypeError: can't compare offset-naive and offset-aware datetimes`,
    which surfaces as a 500 on `/api/auth/refresh`. Returning a naive UTC
    `datetime` for every comparison (and for values being written to
    naive columns) avoids that mismatch."""
    return datetime.now(timezone.utc).replace(tzinfo=None)


async def create_user(
    db: AsyncSession,
    user_data: UserCreate,
    created_by: Optional[uuid.UUID] = None,
) -> User:
    """Create a new user with a bcrypt-hashed password."""
    user = User(
        username=user_data.username,
        email=user_data.email,
        hashed_password=hash_password(user_data.password),
        full_name=user_data.full_name,
        role=user_data.role,
        created_by=created_by,
    )
    db.add(user)
    await db.commit()
    await db.refresh(user)
    return user


async def authenticate_user(
    db: AsyncSession,
    email: str,
    password: str,
) -> Optional[User]:
    """Authenticate by (email, password). Returns the user on success, None
    otherwise. Implements brute-force lockout: after MAX_FAILED_LOGINS
    consecutive wrong passwords the account is locked for LOCKOUT_DURATION.
    """
    result = await db.execute(select(User).where(User.email == email))
    user = result.scalar_one_or_none()
    if user is None:
        # Constant-time-ish: still run bcrypt against a dummy hash so a
        # timing side-channel can't enumerate which emails exist.
        verify_password(password, "$2b$12$" + "x" * 53)
        return None

    now = _utcnow_naive()

    # Lockout check — short-circuits before bcrypt even runs.
    if user.locked_until and user.locked_until > now:
        return None

    if not verify_password(password, user.hashed_password):
        user.failed_login_attempts = (user.failed_login_attempts or 0) + 1
        if user.failed_login_attempts >= settings.MAX_FAILED_LOGINS:
            user.locked_until = now + timedelta(minutes=settings.LOCKOUT_DURATION_MINUTES)
            user.failed_login_attempts = 0
        await db.commit()
        return None

    # Success: reset counter and stamp last_login.
    user.failed_login_attempts = 0
    user.locked_until = None
    user.last_login = now
    await db.commit()
    return user


async def issue_tokens(
    db: AsyncSession,
    user: User,
    remember_me: bool = False,
) -> tuple[str, str, datetime]:
    """Issue both tokens and persist the refresh token session row.

    Returns (access_token, refresh_token, session_expires_at). The caller
    is responsible for placing the tokens in Set-Cookie headers — see
    `api.auth._set_auth_cookies`.

    When `remember_me` is True the refresh token (and its session row)
    is minted for `REMEMBER_ME_EXPIRE_DAYS` so the mobile client can keep
    the user signed in for up to 30 days without re-authenticating.
    """
    access_token = create_access_token(user.id, user.role)
    expire_days = (
        settings.REMEMBER_ME_EXPIRE_DAYS if remember_me else settings.REFRESH_TOKEN_EXPIRE_DAYS
    )
    refresh_token = create_refresh_token(user.id, user.role, expire_days=expire_days)

    now = _utcnow_naive()
    session_expires_at = now + timedelta(days=expire_days)
    session = UserSession(
        user_id=user.id,
        refresh_token=refresh_token,
        expires_at=session_expires_at,
        is_revoked=False,
        created_at=now,
        last_used=now,
    )
    db.add(session)
    await db.commit()
    return access_token, refresh_token, session_expires_at


async def rotate_refresh_token(
    db: AsyncSession,
    old_refresh_token: str,
) -> Optional[tuple[str, str, User, datetime]]:
    """Validate a refresh token, revoke it, and issue a fresh pair.

    Used by `/auth/refresh`. Returns
    (new_access, new_refresh, user, session_expires_at) on success, or
    None if the token is missing/revoked/expired.

    The new session keeps the same lifetime as the old one, so a
    "remember me" session (30 days) stays 30 days across rotations
    instead of snapping back to the default 7.
    """
    from app.core.security import decode_token

    try:
        payload = decode_token(old_refresh_token, expected_type="refresh")
    except Exception:
        return None

    result = await db.execute(
        select(UserSession).where(
            UserSession.refresh_token == old_refresh_token,
            UserSession.is_revoked == False,  # noqa: E712
        )
    )
    session = result.scalar_one_or_none()
    if session is None:
        return None

    now = _utcnow_naive()
    if session.expires_at <= now:
        return None

    # Inactivity gate: a session that hasn't been touched (rotated / used)
    # within INACTIVITY_TIMEOUT_MINUTES is treated as expired. This makes
    # authentication inactivity-based on top of the fixed expiry.
    if now - session.last_used > timedelta(minutes=settings.INACTIVITY_TIMEOUT_MINUTES):
        return None

    user_result = await db.execute(select(User).where(User.id == session.user_id))
    user = user_result.scalar_one_or_none()
    if user is None:
        return None

    # Preserve the original lifetime (e.g. 30d for remember-me, 7d default).
    lifetime = session.expires_at - session.created_at
    if lifetime <= timedelta(0):
        lifetime = timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)
    new_expires_at = now + lifetime

    # Rotate: revoke the old row, mint a fresh one.
    session.is_revoked = True
    session.last_used = now

    new_access = create_access_token(user.id, user.role)
    new_refresh = create_refresh_token(user.id, user.role, expire_days=lifetime.days)
    db.add(
        UserSession(
            user_id=user.id,
            refresh_token=new_refresh,
            expires_at=new_expires_at,
            is_revoked=False,
            created_at=now,
            last_used=now,
        )
    )
    await db.commit()
    return new_access, new_refresh, user, new_expires_at


async def revoke_session(db: AsyncSession, refresh_token: str) -> bool:
    """Mark a refresh-token session row revoked. Idempotent."""
    result = await db.execute(
        select(UserSession).where(UserSession.refresh_token == refresh_token)
    )
    session = result.scalar_one_or_none()
    if session is None:
        return False
    session.is_revoked = True
    await db.commit()
    return True


async def get_user_by_email(db: AsyncSession, email: str) -> Optional[User]:
    result = await db.execute(select(User).where(User.email == email))
    return result.scalar_one_or_none()


async def get_user_by_id(db: AsyncSession, user_id: uuid.UUID) -> Optional[User]:
    result = await db.execute(select(User).where(User.id == user_id))
    return result.scalar_one_or_none()


async def get_session_by_token(db: AsyncSession, refresh_token: str) -> Optional[UserSession]:
    result = await db.execute(
        select(UserSession).where(
            UserSession.refresh_token == refresh_token,
            UserSession.is_revoked == False,  # noqa: E712 — SQLAlchemy needs the IS FALSE comparison
            UserSession.expires_at > _utcnow_naive(),
        )
    )
    return result.scalar_one_or_none()


# --- Password Reset Methods -------------------------------------------------


async def request_password_reset(db: AsyncSession, email: str) -> bool:
    """
    Request a password reset for the given email.
    
    Always returns True to prevent email enumeration.
    If user exists, generates a code and sends it via email.
    """
    user = await get_user_by_email(db, email)
    if user is None:
        # Don't reveal whether the email exists
        return True

    # Delete any existing reset tokens for this user
    await db.execute(delete(PasswordResetToken).where(PasswordResetToken.user_id == user.id))

    # Generate new code
    code = generate_reset_code()
    expires_at = _utcnow_naive() + timedelta(minutes=15)

    reset_token = PasswordResetToken(
        user_id=user.id,
        code=code,
        attempts=0,
        expires_at=expires_at,
    )
    db.add(reset_token)
    await db.commit()

    # Send email (async, don't wait)
    await send_password_reset_email(email, code)
    return True


async def verify_reset_code(db: AsyncSession, email: str, code: str) -> tuple[bool, str]:
    """
    Verify a password reset code.
    
    Returns (success, message).
    After 3 failed attempts, the token is deleted and user must request a new one.
    """
    user = await get_user_by_email(db, email)
    if user is None:
        return False, "Invalid code."

    result = await db.execute(
        select(PasswordResetToken).where(
            PasswordResetToken.user_id == user.id,
            PasswordResetToken.expires_at > _utcnow_naive(),
        )
    )
    reset_token = result.scalar_one_or_none()

    if reset_token is None:
        return False, "Code expired or not found. Please request a new one."

    if reset_token.code != code:
        reset_token.attempts += 1
        await db.commit()
        
        if reset_token.attempts >= 3:
            await db.delete(reset_token)
            await db.commit()
            return False, "Too many failed attempts. Please request a new code."
        
        return False, f"Invalid code. {3 - reset_token.attempts} attempts remaining."

    # Code is valid
    return True, "Code verified successfully."


async def reset_password(db: AsyncSession, email: str, code: str, new_password: str) -> tuple[bool, str]:
    """
    Reset the user's password using a verified code.
    
    Returns (success, message).
    """
    user = await get_user_by_email(db, email)
    if user is None:
        return False, "Invalid request."

    result = await db.execute(
        select(PasswordResetToken).where(
            PasswordResetToken.user_id == user.id,
            PasswordResetToken.expires_at > _utcnow_naive(),
        )
    )
    reset_token = result.scalar_one_or_none()

    if reset_token is None:
        return False, "Code expired or not found. Please request a new one."

    if reset_token.code != code:
        return False, "Invalid code."

    # Update password
    user.hashed_password = hash_password(new_password)
    user.failed_login_attempts = 0
    user.locked_until = None
    
    # Delete the used reset token
    await db.delete(reset_token)
    await db.commit()
    
    return True, "Password reset successfully."

File: backend-node/src/services/authService.ts
/**
 * User persistence, authentication, and session bookkeeping.
 *
 * Mirrors backend/app/services/auth_service.py. All queries are raw SQL
 * against the existing schema (users, user_sessions, password_reset_tokens);
 * hashing/token logic lives in src/security.ts.
 */
import { randomUUID } from "node:crypto";

import { config } from "../config.js";
import { db } from "../db.js";
import { createAccessToken, createRefreshToken, decodeToken, hashPassword, verifyPassword } from "../security.js";
import { epochUtc, nowIso, nowNaive } from "../serialize.js";
import { generateResetCode, sendPasswordResetEmail } from "./emailService.js";
import { verifyEmailDeliverability } from "./verifaliaService.js";

export interface UserRow {
  id: string;
  username: string;
  email: string;
  hashed_password: string;
  full_name: string | null;
  role: string;
  is_active: boolean;
  is_verified: boolean;
  created_by: string | null;
  created_at: string | null;
  updated_at: string | null;
  last_login: string | null;
  failed_login_attempts: number;
  locked_until: string | null;
  avatar_url: string | null;
  nav_style: string;
  nav_collapsed: boolean;
}

export interface SessionRow {
  id: string;
  user_id: string;
  refresh_token: string;
  expires_at: string;
  is_revoked: boolean;
  created_at: string;
  last_used: string;
}

export interface ResetTokenRow {
  id: string;
  user_id: string;
  code: string;
  attempts: number;
  expires_at: string;
  created_at: string;
}

export interface VerificationRow {
  id: string;
  user_id: string;
  code: string;
  attempts: number;
  expires_at: string;
  created_at: string;
}

/** Look a user up by email first, then by exact username. */
export async function getUserByIdentifier(identifier: string): Promise<UserRow | null> {
  const trimmed = identifier.trim();
  const user = await getUserByEmail(trimmed.toLowerCase());
  if (user) return user;
  return db.get<UserRow>(`SELECT * FROM users WHERE username = $1`, [trimmed]);
}

export async function createUser(
  username: string,
  email: string,
  password: string,
  opts: {
    fullName?: string | null;
    role?: string;
    createdBy?: string | null;
    verified?: boolean;
    avatarUrl?: string | null;
  } = {}
): Promise<UserRow> {
  const id = randomUUID();
  const now = nowIso();
  const role = opts.role ?? "student";
  const createdBy = opts.createdBy ?? null;
  await db.query(
    `INSERT INTO users
       (id, username, email, hashed_password, full_name, role, is_active, is_verified,
        created_by, created_at, updated_at, last_login, failed_login_attempts, locked_until,
        avatar_url, nav_style, nav_collapsed)
     VALUES ($1,$2,$3,$4,$5,$6,TRUE,$7,$8,$9,$9,NULL,0,NULL,$10,'sidebar',FALSE)`,
    [
      id,
      username,
      email,
      hashPassword(password),
      opts.fullName ?? null,
      role,
      opts.verified ? true : false,
      createdBy,
      now,
      opts.avatarUrl ?? null,
    ]
  );
  const user = await getUserById(id);
  if (!user) throw new Error("user insert failed");
  return user;
}

export async function getUserByEmail(email: string): Promise<UserRow | null> {
  return db.get<UserRow>(`SELECT * FROM users WHERE email = $1`, [email]);
}

export async function getUserById(id: string): Promise<UserRow | null> {
  return db.get<UserRow>(`SELECT * FROM users WHERE id = $1`, [id]);
}

async function getSessionByToken(token: string): Promise<SessionRow | null> {
  return db.get<SessionRow>(
    `SELECT * FROM user_sessions WHERE refresh_token = $1 AND is_revoked = FALSE`,
    [token]
  );
}

/** Authenticate by (identifier, password) with brute-force lockout. Returns null on failure. */
export async function authenticateUser(
  identifier: string,
  password: string
): Promise<UserRow | null> {
  const user = await getUserByIdentifier(identifier);
  if (!user) {
    // Constant-time-ish: still run bcrypt against a dummy hash so a timing
    // side-channel can't enumerate which emails exist.
    verifyPassword(password, "$2b$12$" + "x".repeat(53));
    return null;
  }

  const nowEpoch = Date.now();
  const lockedEpoch = epochUtc(user.locked_until);
  if (lockedEpoch !== null && lockedEpoch > nowEpoch) return null;

  if (!verifyPassword(password, user.hashed_password)) {
    const attempts = (user.failed_login_attempts || 0) + 1;
    if (attempts >= config.MAX_FAILED_LOGINS) {
      const lockedUntil = new Date(
        Date.now() + config.LOCKOUT_DURATION_MINUTES * 60_000
      ).toISOString();
      await db.query(
        `UPDATE users SET failed_login_attempts = 0, locked_until = $2 WHERE id = $1`,
        [user.id, lockedUntil]
      );
    } else {
      await db.query(`UPDATE users SET failed_login_attempts = $2 WHERE id = $1`, [
        user.id,
        attempts,
      ]);
    }
    return null;
  }

  await db.query(
    `UPDATE users SET failed_login_attempts = 0, locked_until = NULL, last_login = $2 WHERE id = $1`,
    [user.id, nowIso()]
  );
  return getUserById(user.id);
}

export interface IssuedTokens {
  accessToken: string;
  refreshToken: string;
  sessionExpiresAt: Date;
}

/** Issue both tokens and persist the refresh-token session row. */
export async function issueTokens(user: UserRow, rememberMe = false): Promise<IssuedTokens> {
  const expireDays = rememberMe ? config.REMEMBER_ME_EXPIRE_DAYS : config.REFRESH_TOKEN_EXPIRE_DAYS;
  const accessToken = createAccessToken(user.id, user.role);
  const refreshToken = createRefreshToken(user.id, user.role, expireDays);
  const sessionExpiresAt = new Date(Date.now() + expireDays * 24 * 60 * 60_000);

  await db.query(
    `INSERT INTO user_sessions (id, user_id, refresh_token, expires_at, is_revoked, created_at, last_used)
     VALUES ($1,$2,$3,$4,FALSE,$5,$5)`,
    [randomUUID(), user.id, refreshToken, sessionExpiresAt.toISOString(), nowIso()]
  );
  return { accessToken, refreshToken, sessionExpiresAt };
}

/**
 * Validate a refresh token, revoke it, and issue a fresh pair.
 * Returns null when the token is missing/revoked/expired.
 */
export async function rotateRefreshToken(
  oldRefreshToken: string
): Promise<(IssuedTokens & { user: UserRow }) | null> {
  let payload: { sub: string; role: string };
  try {
    const p = decodeToken(oldRefreshToken, "refresh");
    payload = { sub: p.sub, role: p.role };
  } catch {
    return null;
  }

  const session = await getSessionByToken(oldRefreshToken);
  if (!session) return null;

  const now = Date.now();
  const expiresAt = epochUtc(session.expires_at);
  const lastUsed = epochUtc(session.last_used);
  const createdAt = epochUtc(session.created_at);
  if (expiresAt === null || expiresAt <= now) return null;
  if (now - (lastUsed ?? now) > config.INACTIVITY_TIMEOUT_MINUTES * 60_000) return null;

  const user = await getUserById(session.user_id);
  if (!user) return null;

  const lifetimeMs = (expiresAt - (createdAt ?? expiresAt)) || config.REFRESH_TOKEN_EXPIRE_DAYS * 86_400_000;
  const newExpiresAt = new Date(now + lifetimeMs);

  await db.query(
    `UPDATE user_sessions SET is_revoked = TRUE, last_used = $2 WHERE id = $1`,
    [session.id, nowIso()]
  );
  const newAccess = createAccessToken(user.id, user.role);
  const newRefresh = createRefreshToken(user.id, user.role, Math.max(1, Math.round(lifetimeMs / 86_400_000)));
  await db.query(
    `INSERT INTO user_sessions (id, user_id, refresh_token, expires_at, is_revoked, created_at, last_used)
     VALUES ($1,$2,$3,$4,FALSE,$5,$5)`,
    [randomUUID(), user.id, newRefresh, newExpiresAt.toISOString(), nowIso()]
  );
  return { accessToken: newAccess, refreshToken: newRefresh, sessionExpiresAt: newExpiresAt, user };
}

/** Mark a refresh-token session revoked. Idempotent. */
export async function revokeSession(refreshToken: string): Promise<boolean> {
  const res = await db.query(
    `UPDATE user_sessions SET is_revoked = TRUE WHERE refresh_token = $1 AND is_revoked = FALSE`,
    [refreshToken]
  );
  return (res as unknown as { rowCount: number }).rowCount > 0;
}

// --- password reset -------------------------------------------------------

export async function requestPasswordReset(email: string): Promise<boolean> {
  const user = await getUserByEmail(email);
  if (!user) return true; // never reveal whether the email exists

  // Check deliverability via Verifalia before spending a reset code on a dead
  // mailbox. `null` (unconfigured / API failure) means "unknown" → proceed, so
  // a Verifalia outage can never block a legitimate reset.
  const verifalia = await verifyEmailDeliverability(email);
  if (verifalia && verifalia.classification === "Undeliverable") {
    console.warn(`Verifalia: skipping reset for undeliverable address ${email}`);
    return true; // same generic response — don't leak the classification
  }

  await db.query(`DELETE FROM password_reset_tokens WHERE user_id = $1`, [user.id]);
  const code = generateResetCode();
  const expiresAt = new Date(Date.now() + 15 * 60_000).toISOString();
  await db.query(
    `INSERT INTO password_reset_tokens (id, user_id, code, attempts, expires_at, created_at)
     VALUES ($1,$2,$3,0,$4,$5)`,
    [randomUUID(), user.id, code, expiresAt, nowIso()]
  );
  await sendPasswordResetEmail(email, code);
  return true;
}

export async function verifyResetCode(email: string, code: string): Promise<[boolean, string]> {
  const user = await getUserByEmail(email);
  if (!user) return [false, "Invalid code."];

  const token = await db.get<ResetTokenRow>(
    `SELECT * FROM password_reset_tokens WHERE user_id = $1 AND expires_at > $2 ORDER BY created_at DESC LIMIT 1`,
    [user.id, nowNaive()]
  );
  if (!token) return [false, "Code expired or not found. Please request a new one."];

  if (token.code !== code) {
    const attempts = (token.attempts || 0) + 1;
    if (attempts >= 3) {
      await db.query(`DELETE FROM password_reset_tokens WHERE id = $1`, [token.id]);
      return [false, "Too many failed attempts. Please request a new code."];
    }
    await db.query(`UPDATE password_reset_tokens SET attempts = $2 WHERE id = $1`, [
      token.id,
      attempts,
    ]);
    return [false, `Invalid code. ${3 - attempts} attempts remaining.`];
  }
  return [true, "Code verified successfully."];
}

export async function resetPassword(
  email: string,
  code: string,
  newPassword: string
): Promise<[boolean, string]> {
  const user = await getUserByEmail(email);
  if (!user) return [false, "Invalid request."];

  const token = await db.get<ResetTokenRow>(
    `SELECT * FROM password_reset_tokens WHERE user_id = $1 AND expires_at > $2 ORDER BY created_at DESC LIMIT 1`,
    [user.id, nowNaive()]
  );
  if (!token) return [false, "Code expired or not found. Please request a new one."];
  if (token.code !== code) return [false, "Invalid code."];

  await db.query(
    `UPDATE users SET hashed_password = $2, failed_login_attempts = 0, locked_until = NULL WHERE id = $1`,
    [user.id, hashPassword(newPassword)]
  );
  await db.query(`DELETE FROM password_reset_tokens WHERE id = $1`, [token.id]);
  return [true, "Password reset successfully."];
}

// --- email verification ---------------------------------------------------